src/spec-configuration/containerFeaturesOCI.ts

```diff
@@ -187,34 +208,232 @@ export async function getFeatureBlob(output: Log, env: NodeJS.ProcessEnv, url: s
+        output.write(`Provided feature is not of type 'OCI'.  Cannot publish to an OCI registry.`, LogLevel.Error);
+        return false;
+    }
+
+    const featureRef = featureSet.sourceInformation.featureRef;
+
+    // Generate registry auth token with `pull,push` scopes.
+    const registryAuthToken = await fetchRegistryAuthToken(output, featureRef.registry, featureRef.id, env, 'pull,push');
+    if (!registryAuthToken) {
+        output.write(`Failed to get registry auth token`, LogLevel.Error);
+        return false;
+    }
+
+    // Generate Manifest for given feature artifact.
+    const manifest = await generateCompleteManifest(output, pathToTgz);
+    if (!manifest) {
+        output.write(`Failed to generate manifest for ${featureRef.featureName}`, LogLevel.Error);
+        return false;
+    }
+
+    // If the exact manifest digest already exists in the registry, we don't need to push individual blobs (it's already there!) 
+    const existingFeatureManifest = await fetchOCIFeatureManifestIfExists(output, env, featureRef.id, manifest.digest, registryAuthToken);
+    if (manifest.digest && existingFeatureManifest) {
+        output.write(`Not reuploading blobs, digest already exists.`, LogLevel.Trace);
+        await putManifestWithTags(output, manifest.manifestStr, tags);
+        return true;
+    }
+
+    const blobsToPush = [
+        {
+            name: 'configLayer',
+            digest: manifest.manifestObj.config.digest,
+        },
+        {
+            name: 'tgzLayer',
+            digest: manifest.manifestObj.layers[0].digest,
+        }
+    ];
+
+    // Obtain session ID with `/v2/<namespace>/blobs/uploads/` 
+    const blobPutLocationUriPath = await postUploadSessionId(output, featureRef, registryAuthToken);
+    if (!blobPutLocationUriPath) {
+        output.write(`Failed to get upload session ID`, LogLevel.Error);
+        return false;
+    }
+
+    for await (const blob of blobsToPush) {
+        const { name, digest } = blob;
+        const blobExistsConfigLayer = await checkIfBlobExists(output, featureRef, digest, registryAuthToken);
+        output.write(`blob: ${name} with digest ${digest}  ${blobExistsConfigLayer ? 'already exists' : 'does not exist'} in registry.`, LogLevel.Trace);
+
+        // PUT blobs
+        if (!blobExistsConfigLayer) {
+            putBlob(output, blobPutLocationUriPath, featureRef, digest, registryAuthToken);
+        }
+
+        // Send a final PUT to combine blobs and tag manifest properly.
+        await putManifestWithTags(output, manifest.manifestStr, tags);
+
+    }
+
+    // Success!
+    return true;
+}
+
+// Spec: https://github.com/opencontainers/distribution-spec/blob/main/spec.md#pushing-manifests (PUT /manifests/<ref>)
+export async function putManifestWithTags(output: Log, manifest: string, tags: string[]) {
+    output.write(`Tagging manifest with tags: ${tags.join(', ')}`, LogLevel.Trace);
+    output.write(manifest);
+    throw new Error('Not implemented');
```

Comment:
Should we implement this before merging?

Reply:
yes :) moved to another PR